Gems/ROS2/Code/Include/ROS2/Sensor/Events/EventSourceAdapter.h

```diff
@@ -0,0 +1,190 @@
+                {
+                    editContext
+                        ->Class<EventSourceAdapter<EventSourceT>>(
+                            "Event Source Adapter", "Adapts sensor event source to specified working frequency")
+                        ->ClassElement(AZ::Edit::ClassElements::EditorData, "")
+                        ->ElementAttribute(AZ::Edit::Attributes::AutoExpand, true)
+                        ->DataElement(
+                            AZ::Edit::UIHandlers::Default,
+                            &EventSourceAdapter<EventSourceT>::m_adaptedFrequency,
+                            "Adapted frequency",
+                            "Adapter event signalling frequency")
+                        ->DataElement(
+                            AZ::Edit::UIHandlers::Default,
+                            &EventSourceAdapter<EventSourceT>::m_eventSource,
+                            "Event source configuration",
+                            "Event source configuration")
+                        ->Attribute(AZ::Edit::Attributes::Visibility, AZ::Edit::PropertyVisibility::ShowChildrenOnly);
+                }
+            }
+        }
+
+        //! Activates event source adapter - assigns internal adapted event handler and activates managed event source. In most cases, user
+        //! should call ROS2::EventSourceAdapter::Configure first - this adapter itself would work without this requirement, however
+        //! different event source implementations can behave differently.
+        void Activate()
+        {
+            m_adaptedEventHandler = typename EventSourceT::SourceEventHandlerType(
+                [this](auto... args)
+                {
+                    const AZStd::chrono::duration<float, AZStd::chrono::seconds::period> expectedLoopTime =
+                        ROS2Interface::Get()->GetSimulationClock().GetExpectedSimulationLoopTime();
+
+                    if (IsPublicationDeadline(expectedLoopTime.count()))
+                    {
+                        m_sensorAdaptedEvent.Signal(args...);
+                    }
+                });
+            m_eventSource.ConnectToSourceEvent(m_adaptedEventHandler);
+            m_eventSource.Activate();
+        }
+
+        //! Deactivates event source adapter - deactivates event source and disconnects internal adapted event handler from source event. If
+        //! it will be necessary, this implementation allows multiple consecutive calls of Activate / Deactivate, however user must also
+        //! investigate specific event source implementation with such case in mind.
+        void Deactivate()
+        {
+            m_eventSource.Deactivate();
+            m_adaptedEventHandler.Disconnect();
+        }
+
+        //! Parses sensor configuration into event source configuration.
+        //! @param sensorConfiguration Configuration of sensor using this event source adapter.
+        void Configure(const SensorConfiguration& sensorConfiguration)
```

Comment:
Not sure if anything else than frequency from this configuration would ever be useful at this level.